ui/src/admin/components/DatabaseRow.js

```diff
@@ -0,0 +1,206 @@
+              <input
+                className="form-control"
+                name="name"
+                type="text"
+                defaultValue={duration}
+                placeholder="how long should data last"
+                onKeyDown={(e) => this.handleKeyDown(e, database)}
+                ref={(r) => this.duration = r}
+              />
+            </div>
+          </td>
+          <td>
+            <div className="admin-table--edit-cell">
+              <input
+                className="form-control"
+                name="name"
+                type="number"
+                min="1"
+                defaultValue={replication || 1}
+                placeholder="how many nodes do you have"
+                onKeyDown={(e) => this.handleKeyDown(e, database)}
+                ref={(r) => this.replication = r}
+              />
+            </div>
+         </td>
+          <td className="text-right">
+            <YesNoButtons
+              onConfirm={isNew ? this.handleCreate : this.handleUpdate}
+              onCancel={isNew ? () => onRemove(database, retentionPolicy) : this.handleEndEdit}
+            />
+          </td>
+        </tr>
+      )
+    }
+
+    return (
+      <tr>
+        <td onClick={this.handleStartEdit}>
+          {name}
+          {isDefault ? <span className="default-source-label">default</span> : null}
+        </td>
+        <td onClick={this.handleStartEdit}>{formatRPDuration(duration)}</td>
+        <td onClick={this.handleStartEdit}>{replication}</td>
+        <td className="text-right">
+          <button className="btn btn-xs btn-danger admin-table--delete">
+            {`Delete ${name}`}
+          </button>
+        </td>
+      </tr>
+    )
+  }
+
+  handleClickOutside() {
+    this.handleEndEdit()
+  }
+
+  handleStartEdit() {
+    this.setState({isEditing: true})
+  }
+
+  handleEndEdit() {
+    this.setState({isEditing: false})
+  }
+
+  handleCreate() {
+    const {database, onCreate} = this.props
+    const validInputs = this.getInputValues()
+    if (!validInputs) {
+      return
+    }
+
+    onCreate(database, validInputs)
+    this.handleEndEdit()
+  }
+
+  handleUpdate() {
+    const {database, retentionPolicy, onUpdate} = this.props
+    const validInputs = this.getInputValues()
+    if (!validInputs) {
+      return
+    }
+
+    onUpdate(database, retentionPolicy, validInputs)
+    this.handleEndEdit()
+  }
+
+  handleKeyDown(e) {
+    const {key} = e
+    const {retentionPolicy, database, onRemove} = this.props
+
+
+    if (key === 'Escape') {
+      if (retentionPolicy.isNew) {
+        onRemove(database, retentionPolicy)
+        return
+      }
+
+      this.handleEndEdit()
+    }
+
+    if (key === 'Enter') {
+      if (retentionPolicy.isNew) {
+        this.handleCreate()
+        return
+      }
+
+      this.handleUpdate()
+    }
+  }
+
+  getInputValues() {
+    const name = this.name.value.trim()
+    const duration = this.duration.value.trim()
```

Comment:
I agree!  I'm going to handle the simple case of the input being not empty or below 1 for replication factor.  Then, since the database does a great job of validating these strings I think we should let the database handle the parsing and display DB errors.  This way we don't write the same validation code twice and user errors stay consistent! 😸